pallets/asset/src/lib.rs

```diff
@@ -1042,6 +1038,60 @@ decl_module! {
             Self::deposit_event(RawEvent::PrimaryIssuanceAgentTransfered(did, ticker, old_primary_issuance_agent, None));
             Ok(())
         }
+
+        /// Claim a systematically reserved Polymath Classic (PMC) `ticker`
+        /// and transfer it to the `origin`'s identity.
+        ///
+        /// To verify that the `origin` is in control of the Ethereum account on the books,
+        /// an `ethereum_signature` containing the `origin`'s DID as the message
+        /// must be provided by that Ethereum account.
+        ///
+        /// # Errors
+        /// - `NoSuchClassicTicker` if this is not a systematically reserved PMC ticker.
+        /// - `TickerAlreadyRegistered` if the ticker was already registered, e.g., by `origin`.
+        /// - `TickerRegistrationExpired` if the ticker's registration has expired.
+        /// - `BadOrigin` if not signed.
+        /// - `InvalidEthereumSignature` if the `ethereum_signature` is not valid.
+        /// - `NotAnOwner` if the ethereum account is not the owner of the PMC ticker.
+        #[weight = 250_000_000]
+        pub fn claim_classic_ticker(origin, ticker: Ticker, ethereum_signature: ethereum::EcdsaSignature) -> DispatchResult {
+            // Ensure the ticker is a classic one and fetch details.
+            let ClassicTickerRegistration { eth_owner, .. } = ClassicTickers::get(ticker)
+                .ok_or(Error::<T>::NoSuchClassicTicker)?;
+
+            // Ensure ticker registration is still attached to the systematic DID.
+            let sys_did = SystematicIssuers::ClassicMigration.as_id();
+            match Self::is_ticker_available_or_registered_to(&ticker, sys_did) {
+                TickerRegistrationStatus::RegisteredByOther => return Err(Error::<T>::TickerAlreadyRegistered.into()),
+                TickerRegistrationStatus::Available => return Err(Error::<T>::TickerRegistrationExpired.into()),
+                TickerRegistrationStatus::RegisteredByDid => {}
+            }
+
+            // Ensure we're signed & get did.
+            let sender = ensure_signed(origin)?;
+            let owner_did = Context::current_identity_or::<Identity<T>>(&sender)?;
+
+            // Have the caller prove that they own *some* Ethereum account
+            // by having the signed signature contain the `owner_did`.
+            //
+            // We specifically use `owner_did` rather than `sender` such that
+            // if the signing key's owner DID is changed after the creating
+            // `ethereum_signature`, then the call is rejected
+            // (caller might not have Ethereum account's private key).
+            let eth_signer = ethereum::eth_check(owner_did, b"classic_claim", &ethereum_signature)
+                .ok_or(Error::<T>::InvalidEthereumSignature)?;
+
+            // Now we have an Ethereum account; ensure it's the *right one*.
+            ensure!(eth_signer == eth_owner, Error::<T>::NotAnOwner);
+
```

Comment:
Looking at `_accept_ticker_transfer` we also need to:
```
        <AssetOwnershipRelations>::remove(ticker_details.owner, ticker);
        <AssetOwnershipRelations>::insert(to_did, ticker, AssetOwnershipRelation::TickerOwned);
```

Reply:
@maxsam4 I guess that the `AssetOwnershipRelations` will be wrong once a ticker registration expires. i.e. `is_ticker_available_or_registered_to` would return `Available` even though `AssetOwnershipRelations` shows ownership.

One thing we could do is to treat a registration as still being valid after the expiry, but also allowing anyone else to claim it. i.e. if you register your ticker, and 12 months goes by, provided in the last 6 months (since expiry) no one else has claimed your ticker, your registration is still valid.

This would mean switching the `if` clause order in `is_ticker_available_or_registered_to`:
```
                if now > expiry {
                    // ticker registered to someone but expired and can be registered again
                    return TickerRegistrationStatus::Available;
                } else if ticker_reg.owner == did {
                    // ticker is already registered to provided did (but may expire in future)
                    return TickerRegistrationStatus::RegisteredByDid;
                }
```

Reply:
> Looking at `_accept_ticker_transfer` we also need to:

Done (+ refactored transfer commonalities).

> One thing we could do is to treat a registration as still being valid after the expiry, but also allowing anyone else to claim it. i.e. if you register your ticker, and 12 months goes by, provided in the last 6 months (since expiry) no one else has claimed your ticker, your registration is still valid.

This seems reasonable to me, although the current logic also seems reasonable to me. It's rather orthogonal to this PR though, so I'd suggest doing it as a separate task if we want to change things.

Reply:
Agreed - will wrap it as a separate task - should be a trivial change.

Reply:
I think it's best to keep it as it is and add a dev comment about `TickerRegistrationStatus` being outdated. This is essentially applicable to everything with an Expiry. We don't remove expired data automatically. Anything that can expire must be checked for expiry before being considered valid.